plutus-tx/compiler/Language/PlutusTx/Builtins.hs

```diff
@@ -43,75 +42,106 @@ import           Prelude                   hiding (String, error)
 
 import           Language.PlutusTx.Utils   (mustBeReplaced)
 
+{- Note [Builtin name definitions]
+The builtins here have definitions so they can be used in off-chain code too.
+
+However they *must* be replaced by the compiler when used in Plutus Tx code, so
+in particular they must *not* be inlined, otherwise we can't spot them to replace
+them.
+-}
+
+{-# NOINLINE concatenate #-}
 concatenate :: ByteString -> ByteString -> ByteString
 concatenate = BSL.append
 
+{-# NOINLINE takeByteString #-}
 takeByteString :: Integer -> ByteString -> ByteString
 takeByteString i = BSL.take (fromIntegral i)
 
+{-# NOINLINE dropByteString #-}
 dropByteString :: Integer -> ByteString -> ByteString
 dropByteString i = BSL.drop (fromIntegral i)
 
+{-# NOINLINE emptyByteString #-}
 emptyByteString :: ByteString
 emptyByteString = BSL.empty
 
+{-# NOINLINE sha2_256 #-}
 sha2_256 :: ByteString -> ByteString
 sha2_256 = Hash.sha2
 
+{-# NOINLINE sha3_256 #-}
 sha3_256 :: ByteString -> ByteString
 sha3_256 = Hash.sha3
 
+{-# NOINLINE verifySignature #-}
 verifySignature :: ByteString -> ByteString -> ByteString -> Bool
 verifySignature pubKey message signature =
   fromMaybe False (Crypto.verifySignature pubKey message signature)
 
+{-# NOINLINE equalsByteString #-}
 equalsByteString :: ByteString -> ByteString -> Bool
 equalsByteString = (==)
 
+{-# NOINLINE addInteger #-}
 addInteger :: Integer -> Integer -> Integer
 addInteger = (+)
 
+{-# NOINLINE subtractInteger #-}
 subtractInteger :: Integer -> Integer -> Integer
 subtractInteger = (-)
 
+{-# NOINLINE multiplyInteger #-}
 multiplyInteger :: Integer -> Integer -> Integer
 multiplyInteger = (*)
 
+{-# NOINLINE divideInteger #-}
 divideInteger :: Integer -> Integer -> Integer
 divideInteger = div
 
+{-# NOINLINE remainderInteger #-}
 remainderInteger :: Integer -> Integer -> Integer
 remainderInteger = rem
 
+{-# NOINLINE greaterThanInteger #-}
 greaterThanInteger :: Integer -> Integer -> Bool
 greaterThanInteger = (>)
 
+{-# NOINLINE greaterThanEqInteger #-}
 greaterThanEqInteger :: Integer -> Integer -> Bool
 greaterThanEqInteger = (>=)
 
+{-# NOINLINE lessThanInteger #-}
 lessThanInteger :: Integer -> Integer -> Bool
 lessThanInteger = (<)
 
+{-# NOINLINE lessThanEqInteger #-}
 lessThanEqInteger :: Integer -> Integer -> Bool
 lessThanEqInteger = (<=)
 
+{-# NOINLINE equalsInteger #-}
 equalsInteger :: Integer -> Integer -> Bool
 equalsInteger = (==)
 
+{-# NOINLINE error #-}
 error :: () -> a
-error = mustBeReplaced
+error = mustBeReplaced "error"
 
 -- | An opaque type representing PLC strings.
 data String
```

Comment:
Can we have `newtype String = String Prelude.String` and then actually define everything (`appendString`, `emptyString`, etc), so that those definitions can be used off-chain?

Reply:
The problem is that people often pattern match on `String`, and we can't handle that. We could just die if that happened, but it seems neater to just have an abstract type that people can't manipulate in that way.

(We get away with this for `ByteString` because the constructors aren't exported)

Reply:
But does it contradict what I proposed? We can simply not export the constructor of the newtype, no?

Reply:
Oh I see. Yes, we could do that!